src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,414 @@
+    {
+        /// <summary>
+        /// Rewrite `GetTuple() == (1, 2)` to `tuple.Item1 == 1 &amp;&amp; tuple.Item2 == 2`.
+        /// Also supports the != operator, nullable and nested tuples.
+        ///
+        /// Note that all the side-effects for visible expressions are evaluated first and from left to right. The initialization phase
+        /// contains side-effects for:
+        /// - single elements in tuple literals, like `a` in `(a, ...) == (...)` for example
+        /// - nested expressions that aren't tuple literals, like `GetTuple()` in `(..., GetTuple()) == (..., (..., ...))`
+        /// On the other hand, `Item1` and `Item2` of `GetTuple()` are not saved as part of the initialization phase of `GetTuple() == (..., ...)`
+        ///
+        /// Element-wise conversions occur late, together with the element-wise comparisons. They may not be evaluated.
+        /// </summary>
+        /// There are 2 cases:
+        /// - tuple literal cannot be null and it produces as many temps as elements
+        /// - tuple types or nullable tuple types produce a single temp (and no further temps are initialized)
+        public override BoundNode VisitTupleBinaryOperator(BoundTupleBinaryOperator node)
+        {
+            var boolType = node.Type; // we can re-use the bool type
+            var initialEffectsAndTemps = TupleOperatorSideEffectsAndTemps.GetInstance();
+
+            var returnValue = RewriteTupleOperator(node.Operators, node.Left, node.Right, boolType, initialEffectsAndTemps, node.OperatorKind, inLeftLiteral: true, inRightLiteral: true);
+
+            var (effects, temps) = initialEffectsAndTemps.ToImmutableAndFree();
+            var result = _factory.Sequence(temps, effects, returnValue);
+            return result;
+        }
+
+        private BoundExpression RewriteTupleOperator(TupleBinaryOperatorInfo @operator,
+            BoundExpression left, BoundExpression right, TypeSymbol boolType,
+            TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftLiteral, bool inRightLiteral)
+        {
+            if (@operator.IsSingle())
+            {
+                return RewriteTupleSingleOperator((TupleBinaryOperatorInfo.Single)@operator, left, right, boolType, operatorKind);
+            }
+            else
+            {
+                return RewriteTupleNestedOperators((TupleBinaryOperatorInfo.Multiple)@operator, left, right, boolType, initialEffectsAndTemps, operatorKind, inLeftLiteral, inRightLiteral);
+            }
+        }
+
+        private BoundExpression RewriteTupleNestedOperators(TupleBinaryOperatorInfo.Multiple operators, BoundExpression left, BoundExpression right,
+            TypeSymbol boolType, TupleOperatorSideEffectsAndTemps initialEffectsAndTemps, BinaryOperatorKind operatorKind, bool inLeftLiteral, bool inRightLiteral)
+        {
+            // If either left or right is nullable, produce:
```

Comment:
>If either left or right is nullable, produce: [](start = 15, length = 45)

This feels confusing, it looks to me that either both are nullable, or none is nullable.

Reply:
In `tuple == nullableTuple` and permutations, we can have a mix.

---
In reply to: [171071640](https://github.com/dotnet/roslyn/pull/23356#discussion_r171071640) [](ancestors = 171071640)